sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -911,3 +913,51 @@ object RemoveLiteralFromGroupExpressions extends Rule[LogicalPlan] {
       a.copy(groupingExpressions = newGrouping)
   }
 }
+
+/**
+ * Extracts the equal-join condition if any, so that query planner avoids generating cartsian
+ * product which cause out of memory exception, and performance issues
```

Comment:
Can we make this rule simpler and more general? It seems the rule is aptly named, "CNF factorization", and should just do a transformation on expressions (instead of transforming the plan tree and specific to joins). The rest of the rules should then properly fix the join i think.